### ISSUE TITLE
Add support for deterministic UDF

### ISSUE THREAD

**Issue by panthony** [ClickHouse/ClickHouse#59988]
**Use case**

Be able to rely on Query cache when using deterministic UDF.

As explained by the [QueryCache documentation](https://clickhouse.com/docs/en/operations/query-cache#configuration-settings-and-usage), an UDF is considered not deterministic and as such the Query Cache feature is disabled.

There is a settings `query_cache_nondeterministic_function_handling` to cache it nonetheless but in this case it will also cache non-deterministic result which I do not want.

**Describe the solution you'd like**

In the UDF declaration we should be able to specify if the output function is deterministic or not:

```xml
<functions>
  <function>
     <deterministic>true</deterministic>
  <function>
</functions
```

If yes, then ClickHouse should not opt-out from the Query cache by default.

**Comment by LGDHuaOPER** [comment]
Really good

**Comment by chhetripradeep** [comment]
I would like to pick this task. Thank you!

**Comment by GreetingsEarthling** [comment]
I'm working on this

**Comment by rschu1ze** [comment]
@Ergus ClickHouse's testing infrastructure is described here: https://clickhouse.com/docs/development/tests UDFs can admittedly only be tested with integration tests. tests/integration/test_executable_user_defined_function may be good precedence. The setup of integration tests is a little bit involved, tests/integration/README.md contains the steps to do that.

> I have these preliminary changes: [deterministic_udf](https://github.com/Ergus/ClickHouse/tree/deterministic_udf?rgh-link-date=2025-03-11T22%3A11%3A05.000Z) but I need some tests to confirm that this works. Some suggestion?

That looks like the right direction for sure and I think that this is all what is needed on the C++ side.

> Mi initial idea was to add a non-deterministic function (let's say a random or time) and assert that for consecutive calls with the same arguments I always get the same value to assert that the cache is working. It is tricky, but could work.

Sounds generally reasonable. But to make your life easier you could simply test this instead:
- `SYSTEM DROP QUERY CACHE;`
- run a SELECT which calls a _deterministic_ UDF, make sure to pass `SETTINGS use_query_cache = 1`
- check that `SELECT count(*) FROM system.query_cache;` returns one entry

Then,
- `SYSTEM DROP QUERY CACHE;`
- run a SELECT which calls a _non-deterministic_ UDF, make sure to pass `SETTINGS use_query_cache = 1`
- check that `SELECT count(*) FROM system.query_cache;` returns no entries

**Comment by Ergus** [comment]
Hi @rschu1ze :

When I try this in master I get something a bit unexpected (at least for me):

```
ERGUS :) SYSTEM DROP QUERY CACHE;

SYSTEM DROP QUERY CACHE

Query id: 5ce49848-0b9b-4d9f-8c9e-0851704b573e

Ok.

0 rows in set. Elapsed: 0.002 sec.

ERGUS :) select test_function() SETTINGS use_query_cache=1;

SELECT test_function_nodet()
SETTINGS use_query_cache = 1

Query id: b8a8d7bb-407f-4d5e-9efd-ff5bcc749da4

   ┌─test_function_nodet()─┐
1. │ 7                     │
   └───────────────────────┘

1 row in set. Elapsed: 0.008 sec.

ERGUS :) select count(*) FROM system.query_cache;

SELECT count(*)
FROM system.query_cache

Query id: 0b9bc44a-468f-40d6-ab03-5da11a132966

   ┌─count()─┐
1. │       1 │
   └─────────┘

1 row in set. Elapsed: 0.003 sec.
```

I get this in master, without any change.

FWIK this thread exists because UDF don't use the query_cache by default.

I thought that it could be related with `query_cache_nondeterministic_function_handling` but:

```
SELECT name, value
FROM system.settings
WHERE name = 'query_cache_nondeterministic_function_handling';

SELECT
    name,
    value
FROM system.settings
WHERE name = 'query_cache_nondeterministic_function_handling'

Query id: a389a472-c031-4f73-880d-846e613a9ba6

   ┌─name───────────────────────────────────────────┬─value─┐
1. │ query_cache_nondeterministic_function_handling │ throw │
   └────────────────────────────────────────────────┴───────┘

1 row in set. Elapsed: 0.005 sec. Processed 1.22 thousand rows, 434.40 KB (235.13 thousand rows/s., 83.52 MB/s.)
Peak memory usage: 0.00 B.
```

I launched the server with default options, only added `user_scripts_path` and `user_defined_executable_functions_config` for obvious reasons.

Is this intended or I am missing something here?

**Comment by Ergus** [comment]
Hi Again ;)

I have some time this afternoon to check the issue and so far this is what I found:

-----------------------------------
`astContainsNonDeterministicFunctions` internally uses the visitor code in:

`InDepthNodeVisitor.h:doVisit` which goes throw:
`QuerryResultCache.cpp:HasNonDeterministicFunctionsMatcher:visit`

The problem is that this function uses the UDF function name to call

`FunctionFactory.cpp:FunctionFactory::tryGetImpl`
and search the UDF function by name in the list of registered functions.
`functions` and `case_insensitive_functions`

However, the UDF functions were never registered there (I checked the containers and the Register function with gdb) As the function is not found, the visitor `TryGet` return a `nullptr`

and the function visit has this:
  
```
if (func && !func->isDeterministic())
       data.has_non_deterministic_functions = true;
```

so for UDF the `has_non_deterministic_functions` is never set.

This seems like an issue in master which actually makes me wonder how the first issue was reported if at the moment what I see is that all the UDF are considered deterministic.

-------------------------------------------

Apparently the UDF functions are registered in a different container or need to be registered in the global register.

What is the preferred approach?

**Comment by rschu1ze** [comment]
@Ergus You found an interesting glitch 😄

I can reproduce the behavior that on current `master` and w/o any code changes, UDFs are considered as deterministic because `FunctionFactory` doesn't know about them ...

The same happens for the same reason for UDFs which are defined in SQL, e.g.

```sql
CREATE FUNCTION parity_str AS (n) -> if(n % 2, 'odd', 'even');
SELECT parity_str(4) settings use_query_cache = 1;

SELECT count(*) FROM system.query_cache; -- returns 1
```

(and such UDFs are easier to work with than externally defined UDFs)

UDFs are registered separately in some container in "src/Functions/UserDefined/*" - as far as I see through `IUserDefinedSQLObjectsStorage` (respectively through a derived class). Try setting a breakpoint in `ASTPtr UserDefinedSQLObjectsStorageBase::tryGet(const std::string & object_name) const` and then run my above repro.

**Comment by Ergus** [comment]
I will open a separate issue for this problem and propose a potential fix I have a mind for it. Once that issue is solved I will continue with this one.

**Comment by Ergus** [comment]
Hi again:

On yesterday I completed the changes for this issue, including integration tests.

However, I tried to create query tests, but they seem to fail because the `clickhouse-local` never uses the cache, so the `select count(*) FROM system.query_cache;` returns zero all time.

I debugged it down and it looks like when using `clickhouse-local` the var `can_use_query_result_cache == false` because `settings[Setting::use_query_cache] == false` `.

So:

1. Is this intended or an issue? (I could also look into it if needed ;)

2. If not an issue then the tests I added for #77553 may be partially useless, because they use `clickhouse-local`. So, I will need substitute them with integration tests instead.

**Comment by rschu1ze** [comment]
> However, I tried to create query tests, but they seem to fail because the clickhouse-local never uses the cache, so the select count(*) FROM system.query_cache; returns zero all time.

That's right, the query cache is intentionally disabled for clickhouse-local ([here](https://github.com/ClickHouse/ClickHouse/blob/81d058b4c162da824c20c35d28d6b75a6141ff7c/programs/local/LocalServer.cpp#L817)). The reason is that clickhouse-local runs just a single query - result caching doesn't make sense. I updated the docs [here](https://github.com/ClickHouse/ClickHouse/pull/77684).

> If not an issue then the tests I added for https://github.com/ClickHouse/ClickHouse/issues/77553 may be partially useless, because they use clickhouse-local. So, I will need substitute them with integration tests instead.

Oh, now that you mention this ... indeed clickhouse-local doesn't make sense in the tests of https://github.com/ClickHouse/ClickHouse/pull/77633. Let's discuss over in the PR.

**Comment by Ergus** [comment]
The branch [deterministic_udf](https://github.com/Ergus/ClickHouse/tree/deterministic_udf) with the changes to fix this issue is waiting for [#77633](https://github.com/ClickHouse/ClickHouse/pull/77633). 

It already has the integration tests.
